### PR TITLE
HPCC-22860 Add method for listing available stores

### PR DESCRIPTION
**Pull request by rpastrana**
Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by rpastrana** [comment]
@wangkx please review

File: esp/services/ws_store/ws_storeService.cpp [hunk 1]
```diff
@@ -117,6 +117,37 @@ IEspStore* CwsstoreEx::loadStoreProvider(const char* instanceName, const char* l
     return (IEspStore*) xproc();
 }
 
+bool CwsstoreEx::onListStores(IEspContext &context, IEspListStoresRequest &req, IEspListStoresResponse &resp)
+{
+    const char *user = context.queryUserId();
+    user = "rodrigo";
```
Comment (wangkx) [inline esp/services/ws_store/ws_storeService.cpp]: Why?

File: esp/services/ws_store/ws_storeService.cpp [hunk 2]
```diff
@@ -117,6 +117,37 @@ IEspStore* CwsstoreEx::loadStoreProvider(const char* instanceName, const char* l
...
+    const char * typefilter  = req.getTypeFilter();
+
+    IArrayOf<IEspStoreInfo> storeinfos;
+    Owned<IPropertyTree> stores = m_storeProvider->getStores(namefilter, ownerfilter, typefilter, new CSecureUser(user, nullptr));
+    if (stores)
+    {
+        Owned<IPropertyTreeIterator> iter = stores->getElements("Store");
+        ForEach(*iter)
+        {
+            Owned<IEspStoreInfo> store = createStoreInfo("","");
```
Comment (wangkx) [inline esp/services/ws_store/ws_storeService.cpp]: createStoreInfo() not work?

File: esp/services/ws_store/ws_storeService.cpp [hunk 3]
```diff
@@ -117,6 +117,37 @@ IEspStore* CwsstoreEx::loadStoreProvider(const char* instanceName, const char* l
...
+    const char *user = context.queryUserId();
+    user = "rodrigo";
+    double version = context.getClientVersion();
+
+    const char * namefilter = req.getNameFilter();
+    const char * ownerfilter = req.getOwnerFilter();
+    const char * typefilter  = req.getTypeFilter();
+
+    IArrayOf<IEspStoreInfo> storeinfos;
+    Owned<IPropertyTree> stores = m_storeProvider->getStores(namefilter, ownerfilter, typefilter, new CSecureUser(user, nullptr));
```
Comment (wangkx) [inline esp/services/ws_store/ws_storeService.cpp]: You may fill out the storeinfos inside the m_storeProvider->getStores() unless the getStores() has to return a IPropertyTree in other place.

**Review by wangkx — COMMENTED** [review]
@rpastrana a few minor comments.

File: esp/services/ws_store/ws_storeService.cpp (hunk 3 above)
Comment (rpastrana) [reply esp/services/ws_store/ws_storeService.cpp]: opted against this in order to keep esp generated deps out of the DALIKV library

**Comment by rpastrana** [comment]
@wangkx  please review
@dcamper fyi

**Comment by wangkx** [comment]
@rpastrana looks fine.

**Comment by rpastrana** [comment]
@richardkchapman please merge

**Pull request merged**